### ISSUE TITLE
feat(env): position-constrained action validity (option B from #234)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#251]
## Original Issue

> Option B from architect proposal #234. Per the proposal:
>
> > Currently 10 houses × 2 modes = 20 discrete actions, all reachable from any state. Constraining action validity by position (e.g., only adjacent houses, or k-hop reachable) would shrink the policy space and make spatial structure load-bearing on policy learning rather than just on reward.
>
> The post-#236 results on PR #248 (still cooking) will tell us whether the signal channel alone helps PPO. Regardless of that outcome, action-space constraints are a worth-trying lever because:
> - Smaller per-state action space → tighter gradient signal per choice
> - Spatial proximity becomes a policy constraint (currently only a reward shaping mechanism via `distance_cost_alpha`)
> - Aligns with how most RL benchmarks structure spatial games (Minecraft, NetHack, etc.)
>
> ### Possible designs
> 1. **Adjacent-only**: agent can only target their current house or a directly adjacent one
> 2. **k-hop**: agent can target any house within k ring-distance
> 3. **Continuous reach**: action validity probabilistic based on distance (extension of `positional_default`'s alpha)
>
> ### What to do
> - File design discussion comments on this issue to settle on one of the above
> - Implementation requires Rust core change (action validity mask in `engine/step`), PyO3/WASM action interface update, env action_space change
> - All v1 baselines invariant if validity mask is "always-true"
> - Estimated effort: 8-12 hours
>
> ### References
> - Architect proposal #234 (option B)
> - PR #217 (`positional_default` scenario — provides the spatial-cost groundwork)
> - PR #236 (option A — signal-channel fix)

---

### Recommendation: ship Alternative 2 (k-hop) with `k` parameterized

Adjacent-only (k=1) is the simplest possible mask but is a strict special case of k-hop, and "k=1, 3 valid houses out of 10" is aggressive enough that the env's strategic shape changes a lot. k-hop **subsumes both Alternatives 1 and 3 mathematically** while staying discrete (which preserves the `MultiDiscrete([10,2,2])` shape and avoids the entropy / log-prob plumbing that a continuous probabilistic mask would require):

- k = 1 → adjacent-only (Alternative 1)
- k = 5 → ring-half (everything reachable within half the ring)
- k = 10 (or unset) → fully unconstrained (current behavior, exact backward compat)
- Alternative 3 (continuous reach) is already covered by `positional_default`'s `distance_cost_alpha` — that's a *reward* gradient on distance, which is the soft analog of a hard mask. Layering a probabilistic mask on top would double-count the spatial penalty and tangle two knobs that should stay orthogonal.

So: one new `Scenario` field `action_reach_k: u8` (default `10`, meaning unconstrained = pre-fix). Scenario authors get the full sweep without forking the type system.

**Anchor**: the mask is computed against the agent's `agent_home_positions[i]` (already added in PR #217 for `positional_default`). NOT against the agent's *current* `agent_positions[i]`. This is a deliberate, scenario-design-friendly choice:
- Stable per-agent reachable set → policy can specialize "I am the agent who covers houses 0-2".
- Avoids a recursive feedback loop where agents who get pushed far from home one night become unable to act productively the next night.
- Composes cleanly with `positional_default` (which already pins home positions). For scenarios without explicit home positions, the existing `derive_agent_home_positions` fallback (round-robin: agent `i`'s home = house `i`) applies — see `engine/core.rs:41-64`.

If post-implementation evidence suggests the dynamic-anchor variant (mask anchored at last night's `agent_positions[i]`) is more interesting, that's a one-line follow-up issue with its own `action_anchor: "home" | "current"` knob. Out of scope here.

#### 1. Scenario field (Rust)

In `bucket-brigade-core/src/scenarios.rs` (alongside `distance_cost_alpha`, `distance_metric`):

```rust
/// Issue #251 (option B from #234). Per-agent reachable-house horizon on
/// the 10-ring. An agent at home position `h` can target only houses `j`
/// where `ring_dist_10(h, j) <= action_reach_k`. The default of `10` is
/// the ring's diameter halving constant — i.e. every house is reachable
/// from every home — so pre-#251 scenarios are bit-exactly unchanged.
#[serde(default = "default_action_reach_k")]
pub action_reach_k: u8,

fn default_action_reach_k() -> u8 { 10 }
```

Update `Scenario::validate()` to require `action_reach_k <= 10` (covers the whole ring).

#### 2. Engine enforcement (Rust)

In `bucket-brigade-core/src/engine/core.rs`, before `step` writes anything (between lines 135 and 148), clamp the house index for each agent:

```rust
// Enforce position-constrained action validity (issue #251).
// When action_reach_k < 10, any out-of-reach target is rewritten to
// the agent's home position. This makes the mask a hard floor that
// the policy cannot ignore via gradient-ascent against the reward.
let k = self.scenario.action_reach_k;
let mut sanitized: Vec<Action> = actions.to_vec();
if k < 10 {
    for (agent_idx, action) in sanitized.iter_mut().enumerate() {
        let home = self.agent_home_positions[agent_idx];
        if super::core::ring_dist_10(home, action[0]) > k {
            action[0] = home; // fall back to home position
        }
    }
}
// Then use `&sanitized` (not `actions`) throughout the rest of step().
```

**Open implementation question for Builder**: should out-of-reach targets be *clamped to home* (above) or *clamped to mode=REST* (i.e. `action[1] = 0`, ignore house index)? Clamping to home preserves the policy's house-choice gradient on the boundary; clamping to REST is more honest about "you tried to do something invalid". Curator's lean: clamp to home, because it gives PPO a meaningful gradient at the boundary. Builder should flag if the alternative looks better after a smoke run.

Alternative implementation: expose the mask to the policy (see "Future extension" below) and let the network *sample* only valid actions. This is the textbook MARL approach but requires plumbing the mask through `obs` → policy → action sampler in `bucket_brigade/training/joint_trainer.py` (~50 LOC). Recommended as a follow-up, not this issue.

#### 3. PyO3 / WASM surface

- `bucket-brigade-core/src/python.rs:90-159`: add `action_reach_k` to the `PyScenario::new` kwargs (with default), getter at the end of the impl block (lines 216-234 pattern).
- `bucket-brigade-core/src/wasm.rs`: no surface change needed — WASM constructs `Scenario` from JSON via serde, so the `#[serde(default)]` handles the optional field.

#### 4. Python side

- `bucket_brigade/envs/scenarios_generated.py`: regenerate via `scripts/generate_python.py` (the codegen path already handles new optional fields — `positional_default` set the precedent).
- `bucket_brigade/envs/bucket_brigade_env.py`: parity sanitization in the pure-Python `step` (~5 lines mirroring the Rust above).
- `definitions/scenarios.json`: optionally add a new `khop_k1` and/or `khop_k2` scenario for empirical work (recommended) — same parameters as `positional_default` plus `action_reach_k: 1` and `action_reach_k: 2`. These are the "interesting" cells for the post-implementation sweep.

#### 5. action_space change

`bucket_brigade/envs/puffer_env_rust.py:135` / `puffer_env.py:80` keep `MultiDiscrete([10, 2, 2])` unchanged — the mask is enforced inside `step`, not in the action space. **This is the key backward-compat design choice.** Policies trained pre-#251 will still load structurally; their out-of-reach actions just get sanitized to home. Gymnasium does not natively support per-state action masks via `action_space`, so the sanitize-in-step approach is the standard idiom (mirrors what PettingZoo, MARL benchmarks do).

### Backward-compat strategy

- **Default `action_reach_k = 10`** → reach covers the whole 10-ring → mask is always-true → engine behavior bit-exact identical to today.
- **All 14 existing scenarios** in `SCENARIOS` (`default`, `easy`, `hard`, `trivial_cooperation`, `early_containment`, `greedy_neighbor`, `sparse_heroics`, `rest_trap`, `chain_reaction`, `deceptive_calm`, `overcrowding`, `mixed_motivation`, `minimal_specialization`, `positional_default`) get `action_reach_k: 10` explicitly via `default_action_reach_k()`. None of them change.
- **All existing baselines** (random, specialist, heuristic agents — see `bucket-brigade-core/src/agents/heuristic.rs`, `bucket_brigade/baselines/`) work unchanged. They emit unconstrained actions; sanitization is a no-op at `k=10`.
- **Trained checkpoints** load structurally because `MultiDiscrete([10, 2, 2])` is unchanged.

### Composition with `positional_default` (PR #217)

`positional_default` already adds `distance_cost_alpha: 0.1` and `agent_home_positions: [0, 3, 5, 8]`. A natural empirical companion scenario for this issue is `positional_default + action_reach_k: 2`, which gives each of the 4 agents a 5-house window (`home ± 2`) on the 10-ring. At k=2, the windows overlap slightly (e.g., agent at home=3 covers {1,2,3,4,5}, agent at home=5 covers {3,4,5,6,7}), which preserves the cooperation requirement on the overlap houses while removing the "parachute to anywhere" pathology.

### Dependency on #239's verdict

Issue #239 (currently `loom:building`) re-derives the post-#236 obs-fix and MAPPO verdicts. Both are still in flight. The cost-benefit math for this issue depends on #239's outcome:

| #239 outcome | Implication for this issue |
|---|---|
| Tier-1 PASS (signaling alone closes gap) | Lower priority — substrate is now PPO-learnable, action-space constraint is a polish lever, not a fix. |
| Tier-2 PARTIAL | This issue's value is **highest** here. Action-space constraint + operative signaling could compose to close the gap. |
| Tier-3 STILL INSUFFICIENT | This issue is one of two next levers (the other is reward shaping #193). Argument is "shrink the policy space" rather than "give the policy more signal". |

**Recommended sequencing**: don't start implementation until #239's verdict is in (1-2 days of compute pending). The implementation here is mechanical, but the empirical sweep that justifies it (3 scenarios × 3 seeds × 100 iters paired vs k=10 baseline) costs ~30 min on `COMPUTE_HOST_PRIMARY` and only pays back if #239 doesn't already resolve the bottleneck.

## Affected Files

- `bucket-brigade-core/src/scenarios.rs` — new `action_reach_k: u8` field with `#[serde(default)]`, `validate()` bounds check, one or two new scenario entries (`khop_k1`, `khop_k2`).
- `bucket-brigade-core/src/engine/core.rs` — sanitize step inputs against the mask before lines 148-152. `ring_dist_10` helper at line 72 is already in place.
- `bucket-brigade-core/src/engine/rewards.rs` — no change (reward computation uses sanitized `actions` already).
- `bucket-brigade-core/src/engine/tests.rs` — new tests: `test_action_reach_k_default_is_unconstrained`, `test_action_reach_k_clamps_out_of_reach_to_home`, `test_action_reach_k_zero_pins_agent_to_home`, `test_khop_scenarios_backward_compat`.
- `bucket-brigade-core/src/python.rs:90-234` — extend `PyScenario::new` signature + add getter (~10 LOC).
- `bucket-brigade-core/src/wasm.rs` — no source change (serde default handles it).
- `bucket_brigade/envs/scenarios_generated.py` — codegen regenerate via `scripts/generate_python.py`.
- `bucket_brigade/envs/bucket_brigade_env.py` — parity sanitization in `step` (~5 LOC); validation in `__post_init__` for the new field.
- `bucket_brigade/envs/puffer_env_rust.py` / `puffer_env.py` — no change (action space invariant).
- `definitions/scenarios.json` — add `khop_k1`, `khop_k2` entries (optional but recommended for the empirical sweep).
- `tests/test_environment.py` — Python parity test class `TestActionReachK` (~6 tests mirroring the Rust suite).
- `tests/test_code_generation.py` — bump per-field consistency check to recognize `action_reach_k`.
- `web/src/utils/scenarioGenerator.generated.ts` — codegen regenerate.
- `scripts/generate_python.py` — extend the emit-only-when-set logic for the new field (one-line addition).

## Test Plan

- [ ] Manual verification: Construct `BucketBrigade::new` with `action_reach_k: 1` and 4 agents at `[0, 3, 5, 8]`. Feed an out-of-reach action (e.g., agent 0 targeting house 7). Verify agent 0's actual target is rewritten to house 0 (its home) and its work cost reflects the home target, not house 7.
- [ ] Backward compat: every test in `bucket-brigade-core/src/engine/tests.rs` and `bucket-brigade-core/src/scenarios.rs::tests::` still passes with `action_reach_k` defaulted to 10 (33/33 expected).
- [ ] New Rust unit tests (4 new): default-unconstrained, clamps-to-home, k=0 pins agent to home, JSON deserialize round-trip omitting the field.
- [ ] New Python parity tests (4-6 new) in `tests/test_environment.py::TestActionReachK` mirroring the Rust suite.
- [ ] Cross-language consistency: `tests/test_code_generation.py::test_python_scenarios_match_json` and `test_scenario_count_consistency` pass after JSON updates.
- [ ] Pre-flight smoke (on `COMPUTE_HOST_PRIMARY`, post-implementation): single 1-iter PPO cell on `khop_k2` scenario confirms `joint_trainer.py` doesn't crash and saved checkpoints load.
- [ ] Edge cases:
  - `action_reach_k: 0` (every agent pinned to home position) — useful as a sanity check.
  - `action_reach_k: 10` exactly matches pre-#251 outputs bit-for-bit (deterministic seed comparison).
  - `action_reach_k: 5` on a 4-agent scenario with home positions `[0, 3, 5, 8]` — verify every house is reachable from at least one agent (no orphan houses).
- [ ] Empirical sweep (separate research follow-up, post-merge): paired comparison `default` vs `default + action_reach_k=2`, 3 seeds × 100 iters, on `COMPUTE_HOST_PRIMARY`. Hooks into existing `experiments/p3_specialization/train.py` machinery.

## Dependencies

- [ ] #239 — Post-#236 obs-fix + MAPPO re-validation (`loom:building`). Verdict shapes whether this issue is high-priority or low-priority; see decision matrix above. Implementation here is **not gated** on #239, but the empirical sweep that justifies merging is.

This issue can be implemented and merged independently of #239; the empirical justification is contingent on #239's verdict.

## Effort refinement

Original estimate: 8-12 hours. Curator-revised estimate after exploring the codebase:

| Component | Hours |
|---|---|
| Rust scenario field + validation + tests | 1.5 |
| Engine sanitization + tests | 1.0 |
| PyO3 surface | 0.5 |
| Python parity (`bucket_brigade_env.py`) + tests | 1.0 |
| Codegen regen (Python + TS) | 0.5 |
| JSON scenario additions (`khop_k1`, `khop_k2`) | 0.5 |
| End-to-end smoke (manual on remote) | 0.5 |
| PR plumbing + Judge cycle | 1.5 |
| **Subtotal (implementation)** | **7.0** |
| Empirical sweep (paired comparison, ~3 cells × 100 iters) | 1.5 (compute) + 1.0 (analysis) |
| **Total** | **~9.5 hours** |

Implementation alone is **~7 hours**, fitting comfortably in a single Builder PR (well under the 6h / 8 files / 400 LOC scope-guard if the JSON scenario additions are split off into a follow-up).

## Future extension (out of scope)

Per-state observable action mask: extend `AgentObservation` to include a `valid_actions: [bool; 10]` mask field. Plumb through to `joint_trainer.py` so PPO samples from the masked distribution (`logits[~mask] = -inf` before softmax). This is the textbook MARL action-masking idiom and would be **strictly better** than the sanitize-in-step approach for policy learning (no gradient wasted on invalid choices). Scope: ~50 additional LOC across `lib.rs`, `engine/observation.rs`, `python.rs`, `joint_trainer.py`, `networks.py`. File as a follow-up issue if this implementation's empirical results are promising.

## References

- Architect proposal #234 (option B — this issue)
- PR #217 (`positional_default` scenario — spatial-cost groundwork; provides `agent_home_positions` field this issue's mask is anchored to)
- PR #236 (option A — signal-channel fix; orthogonal to this issue, composes cleanly)
- Issue #239 (post-#236 re-validation; verdict shapes priority — see Dependencies)
- Issue #193 (reward shaping — alternative next-lever if signaling-alone insufficient)
- `bucket-brigade-core/src/engine/core.rs:72` — existing `ring_dist_10` helper
- `bucket-brigade-core/src/engine/core.rs:41-64` — existing `derive_agent_home_positions` fallback

**Comment by rjwalters** [comment]
Curating this issue per shepherd request.

Curator pass complete (2026-05-16). Highlights:
- **Design recommendation**: Alternative 2 (k-hop) parameterized by a new `Scenario.action_reach_k: u8` field (default 10 = unconstrained = pre-fix behavior). k-hop mathematically subsumes Alternative 1 (k=1) and renders Alternative 3 redundant with the existing `distance_cost_alpha` knob from PR #217. Mask is anchored at `agent_home_positions[i]` to compose cleanly with `positional_default`.
- **Action space invariant**: `MultiDiscrete([10, 2, 2])` stays unchanged; the mask is enforced inside `engine/step` (sanitize out-of-reach targets to home), so pre-#251 trained checkpoints still load.
- **Backward compat**: serde-defaulted `action_reach_k = 10` means all 14 existing scenarios are bit-exactly identical to today.
- **Effort revised**: 7h implementation (down from 8-12 estimate). Empirical sweep adds ~2.5h on remote.
- **Dependency call-out**: #239 (post-#236 obs-fix + MAPPO re-validation) is currently `loom:building`. Implementation here is not gated on it, but the empirical justification is — the verdict matrix in the body explains the three branches.
- **Future extension** (filed as a separate issue if results warrant): observable per-state action mask plumbed to PPO via `AgentObservation` and `joint_trainer.py`'s sampler. Textbook MARL idiom, ~50 LOC.

Added: implementation sketch with file:line citations, backward-compat strategy, Affected Files (11 files), Test Plan (10 checkboxes covering Rust/Python/codegen/edge cases/empirical sweep), Dependencies, Effort breakdown, Future Extension scope. Original issue text preserved.
